Introducing language.md (#2712)
ILIAS-eLearning/ILIAS: docs/development/language.md

ILIAS Language Handling
=======================

> ILIAS offers multi-language support for the interface of ILIAS.

Reviewer: I suggest to explicitly use `user interface` (and maybe `exported documents`) here, because ILIAS also provides other interfaces, e.g. CLI, where multi-language support is not offered.

Author: You're right. Accepted.

# Guidelines
1.  All language entries are text strings and stored in language files in the 
    subdirectory `/lang`. Each language entry has the format:

        `language_module_ID#:#variable_ID#:#text_content###comment`

>     The elements `language_module_ID`, `variable_ID` and `variable_ID`  are required, 

Reviewer: `variable_ID` is listed twice here.

Reviewer 2: @mjansenDatabay : Thank you for the carefull reading! This was an oversight, it should be
```suggestion
    The elements `language_module_ID`, `variable_ID` and `text_content` MUST be set, 
```

>     The elements `language_module_ID`, `variable_ID` and `variable_ID`  are required, 

Reviewer: Maybe we should use the capitalized keywords `MUST`, `SHOULD`, `REQUIRED`, `OPTIONAL` etc., as desribed in [RFC2199](https://www.ietf.org/rfc/rfc2119.txt) (and already/mostly used below).

Reviewer 2: Agreed, see comment above.

    while `comment` is optional and can be used for additional information about an 
    entry, e.g. „`07 02 2020 new variable`“.
    
2. The `variable_ID` of a language entry MUST be unique within the whole language file. This avoids conflicts in the 
presentation of language entries because the `language_module_ID` is not taken into consideration when ILIAS inserts 

> language entries into the HTML output.

Reviewer: IMO we should remove `HTML` here, because the translations are also use in andy kind of export (PDF, CSV, Excel, ...).

Reviewer 2: I would tend to agree:
```suggestion
language entries into the output.
```

Author: I fully agree.

`common#:#`. The language_module_ID `common` MUST only be used for language entries that are used by various 
components and in combined contexts. Because this block is always read from the database into memory for each 
request, the use of new `common` variables SHOULD be minimised.
 
5. To keep the language files maintainable and facilitate translation and creation of new language versions, 
the amount of language entries should be as low as possible. Therefore, language entries that are no longer 
used in ILIAS due to refactorings or changes in the code MUST be removed from the English language file. 
Whenever possible, existing language entries SHOULD be reused and probably moved to the 'common' module to 
avoid multiple entries of the same meaning.

6. The English language file is the master language file. New variables MUST be added at least to this file, 
since we synchronise the variables when preparing a new ILIAS release. If a variable exists in a file of another 
language but not in the English one, the entry will be removed from the file.

# Additional Information

> Adding new entries into language files will not make them available automatically in the user interface. You need to 

Reviewer: I suggest:

> Adding new entries into language files will not make them automatically available in the user interface. You need to refresh the languages by executing the `Refresh Languages` action in the global ILIAS language administration (`Administration » Languages`).

Author: Thanks for your review, Michael. All suggestions accepted!

refresh the languages by executing Administration » Languages » Refresh Languages in your ILIAS installation.
 

> The global language object `$lng`, an instance of class `ilLanguage`, provides methods to access these strings in the 

Reviewer: I suggrest: 

> The global language object can be retrieved from the dependency injection container by using/calling `$DIC['lng']` or `$DIC->language()`. It is an instance of class `ilLanguage` and provides methods to access these strings in the language of the user in the current authentication context. This is done by using the functions `loadLanguageModule()` and `txt()`.

Author: I fully agree.

language of the user currently logged in. This is done by using the functions `loadLanguageModule()` and `txt()`.

        $lng->loadLanguageModule("frm");
        $tpl->setVariable("TEXT", $lng->txt("frm_new_posting"));    

# Roles
The language handling process in ILIAS knows four distinct roles:

* The *First language maintainer* is managing the language handling process and fixing or delegating language bugs. The
first maintainer has to be notified about newly introduced languages and changes of the language version maintainers.
* The *Second language maintainer* is responsible for the language component.

> * *Developers* are adding new language entries to their components, modifying them or remove them from the code.

Reviewer: I am certainly not THE English expert, but if we use `are adding` and `modifying` here, we also should use `removing`, right?

Reviewer 2: I would suggest to use the simpler form here:
```suggestion
* *Developers* add new language entries to their components, modify them, or remove them from the code.
```

Author: Fine for me

> * *Language version maintainers* are volunteers and take care of a specific language version and translating new entries 

Reviewer: Same here, maybe `taking care` and `translating`?

Reviewer 2: Same here:
```suggestion
* *Language version maintainers* are volunteers, take care of a specific language version and translate new entries.
```

Author: Fully agree!
